pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/builder/BaseRoutingTableBuilder.java

```diff
@@ -86,10 +148,52 @@ protected void setRoutingTables(List<Map<String, List<String>>> routingTables) {
     return _routingTables;
   }
 
-  protected void handleNoServingHost(String segmentName) {
-    LOGGER.error("Found no server hosting segment {} for table {}", segmentName, _tableName);
-    if (_brokerMetrics != null) {
-      _brokerMetrics.addMeteredTableValue(_tableName, BrokerMeter.NO_SERVING_HOST_FOR_SEGMENT, 1);
+  /**
+   * Computes a routing table on-the-fly using a mapping of segment to servers. Because of the performance concern,
+   * the default behavior is to randomly pick a server among available servers for each segment.
+   *
+   * @param segmentsToQuery a list of segments that need to be processed for a particular query
+   * @return a routing table
+   */
+  public Map<String, List<String>> computeDynamicRoutingTable(Set<String> segmentsToQuery) {
```

Comment:
Since _segmentToServersMap can be changed anytime, let's pass both segmentToServersMap and segmentsToQuery as parameters.

Reply:
done